crates/vhost-device-gpio/src/backend.rs

```diff
@@ -56,30 +58,59 @@ struct GpioArgs {
     #[clap(short = 'c', long, default_value_t = 1)]
     socket_count: usize,
 
-    /// List of GPIO devices, one for each guest, in the format <N1>[:<N2>]. The first entry is for
-    /// Guest that connects to socket 0, next one for socket 1, and so on. Each device number here
-    /// will be used to access the corresponding /dev/gpiochipX. Example, "-c 4 -l 3:4:6:1"
+    /// List of GPIO devices, one for each guest, in the format
+    /// <N1>[:<N2>]. The first entry is for Guest that connects to
+    /// socket 0, next one for socket 1, and so on. Each device number
+    /// here will be used to access the corresponding /dev/gpiochipX.
+    /// or simulate a GPIO device with N pins. Example, "-c 4 -l
+    /// 3:s4:6:s1"
     #[clap(short = 'l', long)]
     device_list: String,
 }
 
+type GpioChipID = u32;
+type NumGpios = u32;
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+enum GpioDeviceType {
+    PhysicalDevice(GpioChipID),
+    SimulatedDevice(NumGpios),
+}
+
+impl GpioDeviceType {
+    fn new(cfg: &str) -> Result<Self> {
+        match cfg.strip_prefix('s') {
+            Some(num) => {
+                let num_gpios = num.parse::<u32>().map_err(Error::ParseFailure)?;
+                Ok(GpioDeviceType::SimulatedDevice(num_gpios))
+            }
+            _ => {
+                let devid = cfg.parse::<u32>().map_err(Error::ParseFailure)?;
+                Ok(GpioDeviceType::PhysicalDevice(devid))
+            }
+        }
+    }
+}
+
 #[derive(Debug, PartialEq)]
 pub(crate) struct DeviceConfig {
-    inner: Vec<u32>,
+    inner: Vec<GpioDeviceType>,
 }
 
 impl DeviceConfig {
     fn new() -> Self {
         Self { inner: Vec::new() }
     }
 
-    fn contains_device(&self, number: u32) -> bool {
-        self.inner.iter().any(|elem| *elem == number)
+    fn contains_device(&self, device: GpioDeviceType) -> bool {
+        self.inner.contains(&device)
     }
 
-    fn push(&mut self, device: u32) -> Result<()> {
-        if self.contains_device(device) {
-            return Err(Error::DeviceDuplicate(device));
+    fn push(&mut self, device: GpioDeviceType) -> Result<()> {
+        if let GpioDeviceType::PhysicalDevice(id) = device {
+            if self.contains_device(device) {
+                return Err(Error::DeviceDuplicate(id));
+            }
```

Comment:
Not sure I follow what you want here.

Reply:
It was not a particularly strong suggestion, just a wild idea to avoid the `[cfg]` on the other match arm. But thinking about it, it would be hard to implement a meaningful equality for the SimulatedDevice. So probably it was a stupid suggestion from the beginning.